File: core/expr/expr_test.go
package expr

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// ═══════════════════════════════════════════════════════════════
// The core thesis: the SAME automation rule, expressed by different
// platforms, should produce structurally equivalent Expr trees.
//
// "Temperature above 25°C → turn on AC"
//
// Tuya:  conditions[0].entity_type=1, dp_id="4", comparator=">", value=25
// HA:    trigger: numeric_state, entity_id: sensor.temp, above: 25
// ST:    if.greaterThan(device.main.temperatureMeasurement.temperature, 25)
// Homey: trigger.uri: homey:device:sensor, id: measure_temperature
//
// All map to: Gt(State("sensor", "temperature"), Lit(25))
// ═══════════════════════════════════════════════════════════════

func TestCrossPlatformEquivalence_TemperatureThreshold(t *testing.T) {
	// Tuya representation: dp_id "4" is temperature, comparator ">", value 25
	tuya := GtExpr(
		State("tuya-device-001", "temperature"),
		Lit(25),
	)

	// HA representation: sensor.temperature above 25
	ha := GtExpr(
		State("sensor.living_room_temp", "temperature"),
		Lit(25),
	)

	// SmartThings representation
	st := GtExpr(
		StateWithComponent("st-device-001", "main", "temperature"),
		Lit(25),
	)

	// All three are structurally equivalent
	assert.True(t, StructuralEquiv(tuya, ha), "Tuya ≡ HA (structure)")
	assert.True(t, StructuralEquiv(ha, st), "HA ≡ ST (structure)")

	// Semantic equivalence (attribute matches, device IDs differ)
	assert.True(t, Equiv(tuya, ha), "Tuya ≡ HA (semantic)")
	assert.True(t, Equiv(ha, st), "HA ≡ ST (semantic)")
}

func TestCrossPlatformEquivalence_MotionLight(t *testing.T) {
	// "When motion detected AND light is off → turn on light, wait 5min, turn off"
	//
	// This is the most common IoT automation pattern across ALL platforms.

	// The trigger: motion == true
	motionTrigger := EqExpr(State("sensor-01", "motion"), Lit(true))

	// The condition: light == off
	lightOff := EqExpr(State("light-01", "state"), Lit("off"))

	// The action sequence: [turn_on, delay 300s, turn_off]
	actions := SeqExpr(
		Cmd("light-01", "state", "on"),
		DelayExpr(300),
		Cmd("light-01", "state", "off"),
	)

	// Full automation expression
	auto := AndExpr(motionTrigger, lightOff)

	assert.True(t, IsValid(motionTrigger))
	assert.True(t, IsValid(lightOff))
	assert.True(t, IsValid(actions))
	assert.True(t, IsValid(auto))

	// Device refs
	refs := DeviceRefs(auto)
	assert.Contains(t, refs, "sensor-01")
	assert.Contains(t, refs, "light-01")

	actionRefs := DeviceRefs(actions)
	assert.Contains(t, actionRefs, "light-01")
	assert.Len(t, actionRefs, 1) // Only light-01
}

func TestCombinator_CompoundConditions(t *testing.T) {
	// "(temp > 25 OR humidity > 80) AND time_between(22:00, 06:00)"
	tempHigh := GtExpr(State("sensor-01", "temperature"), Lit(25))
	humidHigh := GtExpr(State("sensor-01", "humidity"), Lit(80))

	timeCond := BetweenExpr(Time("now"), Lit("22:00"), Lit("06:00"))

	full := AndExpr(OrExpr(tempHigh, humidHigh), timeCond)

	assert.True(t, IsValid(full))
	assert.Equal(t, 4, Depth(full))    // And → Or → Gt → StateRef
	assert.Equal(t, 12, NodeCount(full)) // And(1) + Or(1)+Gt(3)+Gt(3) + Between(1)+Time(1)+Lit(1)+Lit(1)

	refs := DeviceRefs(full)
	assert.Len(t, refs, 1) // Only sensor-01
	assert.Contains(t, refs, "sensor-01")
}

func TestNot_InvertedCondition(t *testing.T) {
	// Homey's "inverted: true" on a condition
	// Light is NOT on → NotExpr(EqExpr(State, Lit("on")))
	lightOn := EqExpr(State("light-01", "state"), Lit("on"))
	lightNotOn := NotExpr(lightOn)

	assert.True(t, IsValid(lightNotOn))
	assert.False(t, Equiv(lightOn, lightNotOn))

	// Not(Not(x)) is structurally deeper but same idea
	doubleNot := NotExpr(NotExpr(lightOn))
	assert.True(t, IsValid(doubleNot))
	assert.Equal(t, 4, Depth(doubleNot))
}

func TestSeq_ActionSequence(t *testing.T) {
	// Turn on → delay 5s → set brightness 80% → notify
	actions := SeqExpr(
		Cmd("light-01", "state", "on"),
		DelayExpr(5),
		Cmd("light-01", "brightness", 80).WithMeta("service", "light.turn_on"),
		NotifyExpr("Light is on!"),
	)

	assert.True(t, IsValid(actions))
	assert.Equal(t, 4, len(actions.Children))

	// Seq is order-dependent
	reversed := SeqExpr(
		NotifyExpr("Light is on!"),
		Cmd("light-01", "brightness", 80),
		DelayExpr(5),
		Cmd("light-01", "state", "on"),
	)
	assert.False(t, Equiv(actions, reversed), "Seq order matters")
}

func TestParallel_OrderIndependent(t *testing.T) {
	a := Cmd("light-01", "state", "on")
	b := Cmd("light-02", "state", "on")
	c := Cmd("light-03", "state", "on")

	p1 := ParallelExpr(a, b, c)
	p2 := ParallelExpr(c, a, b) // Different order

	assert.True(t, Equiv(p1, p2), "Parallel is order-independent")
}

func TestValidation_Errors(t *testing.T) {
	tests := []struct {
		name string
		expr *Expr
		ok   bool
	}{
		{"valid eq", EqExpr(State("d", "s"), Lit(1)), true},
		{"eq needs 2 children", &Expr{Op: Eq, Children: []*Expr{Lit(1)}}, false},
		{"between needs 3", BetweenExpr(Lit(1), Lit(2), Lit(3)), true},
		{"between with 2", &Expr{Op: Between, Children: []*Expr{Lit(1), Lit(2)}}, false},
		{"not needs 1", NotExpr(Lit(true)), true},
		{"not with 2", &Expr{Op: Not, Children: []*Expr{Lit(1), Lit(2)}}, false},
		{"and needs 2+", AndExpr(Lit(1), Lit(2)), true},
		{"and with 1", &Expr{Op: And, Children: []*Expr{Lit(1)}}, false},
		{"state_ref without ref", &Expr{Op: StateRef}, false},
		{"state_ref without deviceID", &Expr{Op: StateRef, Ref: &DeviceRef{Attribute: "temp"}}, false},
		{"delay without value", &Expr{Op: Delay}, false},
		{"command without ref", &Expr{Op: Command, Value: "on"}, false},
		{"literal is leaf", Lit(42), true},
		{"literal with children", &Expr{Op: Literal, Value: 42, Children: []*Expr{Lit(1)}}, false},
		{"in needs 2+", InExpr(Lit(1), Lit(2), Lit(3)), true},
		{"in needs at least 2", &Expr{Op: In, Children: []*Expr{Lit(1)}}, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			errs := Validate(tt.expr)
			if tt.ok {
				assert.Empty(t, errs, "expected valid")
			} else {
				assert.NotEmpty(t, errs, "expected errors")
			}
		})
	}
}

func TestEquiv_AndIsOrderIndependent(t *testing.T) {
	a := EqExpr(State("d1", "temp"), Lit(25))
	b := EqExpr(State("d2", "humidity"), Lit(60))

	ab := AndExpr(a, b)
	ba := AndExpr(b, a)

	assert.True(t, Equiv(ab, ba), "AND is commutative")
}

func TestEquiv_DifferentOpsNotEqual(t *testing.T) {
	a := GtExpr(State("d", "temp"), Lit(25))
	b := LtExpr(State("d", "temp"), Lit(25))

	assert.False(t, Equiv(a, b))
}

func TestEquivWithMapping(t *testing.T) {
	// Tuya uses true/false for on/off, HA uses "on"/"off"
	tuya := EqExpr(State("device-01", "switch"), Lit(true))
	ha := EqExpr(State("switch.lamp", "switch"), Lit("on"))

	// Without mapping: not equivalent (true != "on")
	assert.False(t, Equiv(tuya, ha))

	// With mapping: true → "on", false → "off"
	mapper := func(ref *DeviceRef, value any) any {
		switch v := value.(type) {
		case bool:
			if v {
				return "on"
			}
			return "off"
		}
		return value
	}

	assert.True(t, EquivWithMapping(tuya, ha, mapper), "with mapper: true == on")
}

func TestJSON_RoundTrip(t *testing.T) {
	// Complex expression should survive JSON round-trip
	expr := AndExpr(
		GtExpr(State("sensor-01", "temperature"), Lit(25)),
		NotExpr(EqExpr(State("light-01", "state"), Lit("on"))),
	)

	data, err := json.Marshal(expr)
	require.NoError(t, err)

	var restored Expr
	err = json.Unmarshal(data, &restored)
	require.NoError(t, err)

	assert.Equal(t, And, restored.Op)
	assert.Len(t, restored.Children, 2)
	assert.Equal(t, Gt, restored.Children[0].Op)
	assert.Equal(t, Not, restored.Children[1].Op)

	t.Logf("JSON: %s", string(data))
}

func TestMeta_PlatformHints(t *testing.T) {
	// Meta preserves platform-specific info without polluting the tree
	cmd := Cmd("light-01", "state", "on").
		WithMeta("ha_service", "light.turn_on").
		WithMeta("tuya_dp_id", "1").
		WithMeta("st_capability", "switch")

	assert.True(t, IsValid(cmd))
	assert.Equal(t, "light.turn_on", cmd.Meta["ha_service"])
	assert.Equal(t, "1", cmd.Meta["tuya_dp_id"])
	assert.Equal(t, "switch", cmd.Meta["st_capability"])
}

func TestSceneExpr(t *testing.T) {
	scene := SceneExpr("night-mode-001")
	assert.True(t, IsValid(scene))
	assert.True(t, scene.IsAction())
	assert.Equal(t, "night-mode-001", scene.Value)
}

func TestDepthAndNodeCount(t *testing.T) {
	// Simple: Eq(State, Lit) → depth 2, nodes 3
	simple := EqExpr(State("d", "s"), Lit(1))
	assert.Equal(t, 2, Depth(simple))
	assert.Equal(t, 3, NodeCount(simple))

	// Nested: And(Eq(State,Lit), Or(Gt(State,Lit), Lt(State,Lit)))
	nested := AndExpr(
		EqExpr(State("d", "s"), Lit(1)),
		OrExpr(
			GtExpr(State("d", "t"), Lit(2)),
			LtExpr(State("d", "u"), Lit(3)),
		),
	)
	assert.Equal(t, 4, Depth(nested))
	assert.Equal(t, 11, NodeCount(nested))
}

File: semantic/test/iot/semantic/expr_test.clj
(ns iot.semantic.expr-test
  "아기방 카메라부터 온도 센서까지 — 실제 IoT 레시피를 Clojure로.

  Go에서 170줄 struct + 생성자가 여기서는 그냥 맵이다.
  테스트가 곧 문서이고, 데이터가 곧 코드다."
  (:require [clojure.test :refer [deftest is testing]]
            [iot.semantic.expr :as e]
            [iot.semantic.equiv :as eq]))

;; ═══════════════════════════════════════════════════════════
;; 1. 데이터가 곧 코드 — S-expression으로 레시피 만들기
;; ═══════════════════════════════════════════════════════════

(deftest expr-is-just-data
  (testing "Expr은 특별한 타입이 아니다. 그냥 맵이다."
    (let [motion (e/eq (e/state-ref "sensor" "motion")
                       (e/lit true))]
      (is (= :eq (:op motion)))
      (is (= 2 (count (:children motion))))
      (is (map? motion) "맵이니까 assoc, dissoc, merge 전부 된다")

      ;; Go에서는 불가능한 것: 맵처럼 다루기
      (is (= :eq (:op (assoc motion :meta {:origin "test"})))
          "아무 키나 붙일 수 있다 — 열린 확장"))))

(deftest recipe-baby-camera
  (testing "아기방 움직임 → 녹화 + 알림 — 안전 최우선 레시피"
    (let [r (e/recipe
             :id "baby_camera_motion"
             :name "아기방 움직임 감지"
             :trigger (e/eq (e/state-ref "camera.baby" "motion")
                            (e/lit true))
             :condition (e/between (e/time-ref "now")
                                   (e/lit "22:00")
                                   (e/lit "06:00"))
             :actions (e/seq-expr
                       (e/cmd "camera.baby" "recording" "start")
                       (e/delay-expr 5)
                       (e/notify "아기방 움직임 감지!")))]

      (is (= "아기방 움직임 감지" (:name r)))
      (is (= :eq (get-in r [:trigger :op])))
      (is (= :seq (get-in r [:actions :op])))
      (is (= 3 (count (get-in r [:actions :children]))))

      ;; 디바이스 참조 수집
      (is (= #{"camera.baby"}
             (e/device-refs (:actions r)))))))

;; ═══════════════════════════════════════════════════════════
;; 2. 교차 플랫폼 동치 — 같은 규칙, 다른 표현
;; ═══════════════════════════════════════════════════════════

(deftest cross-platform-motion-light
  (testing "5개 플랫폼이 '움직임 감지 → 불 켜기'를 다르게 표현한다"
    (let [tuya   (e/eq (e/state-ref "tuya-pir"   "dp_1")           (e/lit true))
          ha     (e/eq (e/state-ref "sensor.pir"  "state")          (e/lit "on"))
          st     (e/eq (e/state-ref "st-pir"      "motion")         (e/lit "active"))
          homey  (e/eq (e/state-ref "homey-pir"   "alarm_motion")   (e/lit true))
          google (e/eq (e/state-ref "google-pir"  "MotionDetection") (e/lit "motionDetected"))
          all    [tuya ha st homey google]]

      (testing "구조적으로 전부 같다: (eq (state-ref ...) (lit ...))"
        (doseq [i (range (count all))
                j (range (inc i) (count all))]
          (is (eq/structural-equiv? (nth all i) (nth all j))
              (str "platform " i " ≡ platform " j)))))))

(deftest commutative-and-or
  (testing "And/Or은 교환법칙 성립 — (and A B) ≡ (and B A)"
    (let [temp-high  (e/gt (e/state-ref "s" "temperature") (e/lit 25))
          humid-high (e/gt (e/state-ref "s" "humidity")    (e/lit 80))
          night      (e/between (e/time-ref "now") (e/lit "22:00") (e/lit "06:00"))

          ;; 플랫폼 A: (temp OR humid) AND night
          expr-a (e/and-expr (e/or-expr temp-high humid-high) night)
          ;; 플랫폼 B: night AND (humid OR temp) — 순서 뒤바뀜
          expr-b (e/and-expr night (e/or-expr humid-high temp-high))]

      (is (eq/equiv? expr-a expr-b)
          "같은 의미다 — 순서만 다르다"))))

(deftest seq-order-matters
  (testing "Seq는 순서가 의미다 — 켜고→끄기 ≠ 끄고→켜기"
    (let [on  (e/cmd "light" "state" "on")
          off (e/cmd "light" "state" "off")
          w   (e/delay-expr 300)]

      (is (not (eq/equiv? (e/seq-expr on w off)
                          (e/seq-expr off w on)))
          "순서가 다르면 의미가 다르다")

      (is (eq/structural-equiv? (e/seq-expr on w off)
                                (e/seq-expr off w on))
          "하지만 구조는 같다 (cmd, delay, cmd)"))))

;; ═══════════════════════════════════════════════════════════
;; 3. 값 매핑 — true ↔ "on" ↔ "active"
;; ═══════════════════════════════════════════════════════════

(deftest value-mapping
  (testing "다른 플랫폼의 값이 같은 의미인지 매핑으로 검증"
    (let [tuya  (e/eq (e/state-ref "dev" "motion") (e/lit true))
          ha    (e/eq (e/state-ref "dev" "motion") (e/lit "on"))
          st    (e/eq (e/state-ref "dev" "motion") (e/lit "active"))

          mapper (fn [_ref v]
                   (case v
                     (true "on" "active") :motion-detected
                     (false "off" "inactive") :no-motion
                     v))]

      (is (not (eq/equiv? tuya ha))
          "매핑 없이는 true ≠ \"on\"")

      (is (eq/equiv? tuya ha mapper)
          "매핑 있으면 true ≡ \"on\"")
      (is (eq/equiv? ha st mapper)
          "\"on\" ≡ \"active\"")
      (is (eq/equiv? tuya st mapper)
          "true ≡ \"active\""))))

;; ═══════════════════════════════════════════════════════════
;; 4. diff — 어디서 갈라졌는가
;; ═══════════════════════════════════════════════════════════

(deftest diff-report
  (testing "equiv?가 false일 때, diff가 '어디서'를 알려준다"
    (let [a (e/and-expr (e/eq (e/state-ref "s" "temp") (e/lit 25))
                        (e/eq (e/state-ref "s" "humid") (e/lit 80)))
          ;; b는 80 대신 90
          b (e/and-expr (e/eq (e/state-ref "s" "temp") (e/lit 25))
                        (e/eq (e/state-ref "s" "humid") (e/lit 90)))
          d (eq/diff a b)]

      (is (not (eq/equiv? a b)))
      (is (seq d) "차이가 있다")
      (is (some #(= :value-mismatch (:type %)) d)
          "값 불일치를 보고한다"))))

;; ═══════════════════════════════════════════════════════════
;; 5. 트리 유틸 — walk, fold, depth
;; ═══════════════════════════════════════════════════════════

(deftest tree-utilities
  (testing "walk-expr: 트리의 모든 노드를 변환"
    (let [expr (e/and-expr (e/eq (e/state-ref "s" "t") (e/lit 25))
                           (e/eq (e/state-ref "s" "h") (e/lit 80)))
          ;; 모든 노드에 :visited 표시
          walked (e/walk-expr #(assoc % :visited true) expr)]
      (is (:visited walked))
      (is (every? :visited (:children walked)))))

  (testing "depth와 node-count"
    (let [simple (e/eq (e/state-ref "s" "t") (e/lit 25))
          nested (e/and-expr simple (e/or-expr simple simple))]
      (is (= 2 (e/depth simple)))    ; eq(1) → state-ref|lit(1)
      (is (= 3 (e/node-count simple)))
      (is (= 4 (e/depth nested)))    ; and → or → eq → leaf
      (is (= 11 (e/node-count nested))))))  ; and(1)+eq(3)+or(1)+eq(3)+eq(3))

;; ═══════════════════════════════════════════════════════════
;; 6. 열린 확장 — 플랫폼 전용 Op
;; ═══════════════════════════════════════════════════════════

(deftest open-extension
  (testing "새 Op은 키워드 하나면 된다 — 코드 수정 없음"
    (let [;; HA의 choose (if-else 분기) — Go의 22종 Op에 없던 것
          ha-choose {:op :ha/choose
                     :children [{:op :eq
                                 :children [(e/state-ref "s" "t") (e/lit "on")]
                                 :then (e/cmd "light" "state" "on")}
                                {:op :eq
                                 :children [(e/state-ref "s" "t") (e/lit "off")]
                                 :then (e/cmd "light" "state" "off")}]}]

      (is (= :ha/choose (:op ha-choose))
          "네임스페이스 키워드로 충돌 없이 확장")

      ;; walk-expr도 그냥 작동한다 — Op을 몰라도 children만 있으면
      (is (= :ha/choose (:op (e/walk-expr identity ha-choose)))
          "walk는 구조만 보고 순회 — Op을 알 필요 없다"))))
